### ISSUE TITLE
Issue: nmap plugin misses HTTP services

### ISSUE THREAD

**Issue by h0ng10** [google/tsunami-security-scanner-plugins#303]
## Issue description
A large part of existing Tsunami plugins identifies vulnerabilities in HTTP based web services. Thus, it is important that these services are identified correctly. Turns out the nmap plugin (or more precise nmap itself) is really bad at this.

For example, I'm currently developing a plugin for a vulnerability in MinIO. MinIO runs a HTTP based API, usually on TCP port 9000. Here the output from nmap (version 7.93), using the same parameters as the nmap plugin:

```
nmap --unprivileged -Pn -n -sT -sV --version-intensity 5 -T4 --script banner 172.21.0.4 -p 9000
Starting Nmap 7.93 ( https://nmap.org ) at 2023-04-06 08:46 CEST
Nmap scan report for 172.21.0.4
Host is up (0.00029s latency).

PORT     STATE SERVICE     VERSION
9000/tcp open  cslistener?
| fingerprint-strings: 
|   GenericLines, Help, Kerberos, RTSPRequest, SIPOptions, SSLSessionReq, TLSSessionReq: 
|     HTTP/1.1 400 Bad Request
|     Content-Type: text/plain; charset=utf-8
|     Connection: close
|     Request
|   GetRequest: 
|     HTTP/1.0 400 Bad Request
|     Accept-Ranges: bytes
|     Content-Length: 232
|     Content-Type: application/xml
|     Server: MinIO
|     Vary: Origin
|     Date: Thu, 06 Apr 2023 06:46:32 GMT
|     <?xml version="1.0" encoding="UTF-8"?>
|     <Error><Code>InvalidRequest</Code><Message>Invalid Request (invalid argument)</Message><Resource>/</Resource><RequestId></RequestId><HostId>ed2576f8-f43d-4c0d-8bf5-60fd0bf780cc</HostId></Error>
|   HTTPOptions: 
|     HTTP/1.0 200 OK
|     Vary: Origin
|     Date: Thu, 06 Apr 2023 06:46:32 GMT
|_    Content-Length: 0
1 service unrecognized despite returning data. If you know the service/version, please submit the following fingerprint at https://nmap.org/cgi-bin/submit.cgi?new-service :
SF-Port9000-TCP:V=7.93%I=5%D=4/6%Time=642E6AC8%P=x86_64-unknown-linux-gnu%
SF:r(GenericLines,67,"HTTP/1\.1\x20400\x20Bad\x20Request\r\nContent-Type:\
SF:x20text/plain;\x20charset=utf-8\r\nConnection:\x20close\r\n\r\n400\x20B
SF:ad\x20Request")%r(GetRequest,190,"HTTP/1\.0\x20400\x20Bad\x20Request\r\
SF:nAccept-Ranges:\x20bytes\r\nContent-Length:\x20232\r\nContent-Type:\x20
SF:application/xml\r\nServer:\x20MinIO\r\nVary:\x20Origin\r\nDate:\x20Thu,
SF:\x2006\x20Apr\x202023\x2006:46:32\x20GMT\r\n\r\n<\?xml\x20version=\"1\.
SF:0\"\x20encoding=\"UTF-8\"\?>\n<Error><Code>InvalidRequest</Code><Messag
SF:e>Invalid\x20Request\x20\(invalid\x20argument\)</Message><Resource>/</R
SF:esource><RequestId></RequestId><HostId>ed2576f8-f43d-4c0d-8bf5-60fd0bf7
SF:80cc</HostId></Error>")%r(HTTPOptions,59,"HTTP/1\.0\x20200\x20OK\r\nVar
SF:y:\x20Origin\r\nDate:\x20Thu,\x2006\x20Apr\x202023\x2006:46:32\x20GMT\r
SF:\nContent-Length:\x200\r\n\r\n")%r(RTSPRequest,67,"HTTP/1\.1\x20400\x20
SF:Bad\x20Request\r\nContent-Type:\x20text/plain;\x20charset=utf-8\r\nConn
SF:ection:\x20close\r\n\r\n400\x20Bad\x20Request")%r(Help,67,"HTTP/1\.1\x2
SF:0400\x20Bad\x20Request\r\nContent-Type:\x20text/plain;\x20charset=utf-8
SF:\r\nConnection:\x20close\r\n\r\n400\x20Bad\x20Request")%r(SSLSessionReq
SF:,67,"HTTP/1\.1\x20400\x20Bad\x20Request\r\nContent-Type:\x20text/plain;
SF:\x20charset=utf-8\r\nConnection:\x20close\r\n\r\n400\x20Bad\x20Request"
SF:)%r(TLSSessionReq,67,"HTTP/1\.1\x20400\x20Bad\x20Request\r\nContent-Typ
SF:e:\x20text/plain;\x20charset=utf-8\r\nConnection:\x20close\r\n\r\n400\x
SF:20Bad\x20Request")%r(Kerberos,67,"HTTP/1\.1\x20400\x20Bad\x20Request\r\
SF:nContent-Type:\x20text/plain;\x20charset=utf-8\r\nConnection:\x20close\
SF:r\n\r\n400\x20Bad\x20Request")%r(SIPOptions,67,"HTTP/1\.1\x20400\x20Bad
SF:\x20Request\r\nContent-Type:\x20text/plain;\x20charset=utf-8\r\nConnect
SF:ion:\x20close\r\n\r\n400\x20Bad\x20Request");

Service detection performed. Please report any incorrect results at https://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 36.58 seconds

```

When the nmap scanner plugin parses the XML output, the [extractServiceFromNmapRun](https://github.com/google/tsunami-security-scanner-plugins/blob/master/google/portscan/nmap/src/main/java/com/google/tsunami/plugins/portscan/nmap/NmapPortScanner.java#L177) method is invoked. To identify web services, it calls Tsunamies  [NetworkServiceUtils.isWebService](https://github.com/google/tsunami-security-scanner/blob/71403e26a2b96d179c171a3b394e88b7499b97ff/common/src/main/java/com/google/tsunami/common/data/NetworkServiceUtils.java#L67) which simply checks the service name against a list of known "nmap patterns":

```java
  private static final ImmutableMap<String, Boolean> IS_PLAIN_HTTP_BY_KNOWN_WEB_SERVICE_NAME =
      ImmutableMap.<String, Boolean>builder()
          .put("http", true)
          .put("http-alt", true) // Some http server are identified as this rather than "http".
          .put("http-proxy", true)
          .put("https", false)
          .put("radan-http", true) // Port 8088, Hadoop Yarn web UI identified as this.
          .put("ssl/http", false)
          .put("ssl/https", false)
          .put("ssl/http-proxy", false)
          .build();
```
This requires that nmap identifies the HTTP service correctly. However, the basic fingerprinting of nmap is less "advanced" as most people assume. To my knowledge the fingerprinting does not allow "dynamic values" (for example the hostname/IP address) in the request. This would be required to send a HTTP 1.1 request. Nmap is therefore only uses a simple "HTTP 1.0" request that does not require a "Host:" header field.

This works fine if the HTTP service still fully supports HTTP 1.0 but for many modern web service implementations, this is no longer the case. The MinIO server for example returns with a HTTP 400 error, which does not match with the service fingerprints from nmap:

```
GET / HTTP/1.0

HTTP/1.0 400 Bad Request
Accept-Ranges: bytes
Content-Length: 232
Content-Type: application/xml
Server: MinIO
Vary: Origin
Date: Thu, 06 Apr 2023 07:06:03 GMT

<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InvalidRequest</Code><Message>Invalid Request (invalid argument)</Message><Resource>/</Resource><RequestId></RequestId><HostId>ed2576f8-f43d-4c0d-8bf5-60fd0bf780cc</HostId></Error>
``` 
As a result, the service is not identified as a web service and is not scanned by plugins that should actually be executed.

## Suggested solution/improvement

I think the best solution would to improve the code in the `extractServiceFromNmapRun`. The code should take the additional scan data from the NMAP output into account and use that to perform additional HTTP / service fingerprinting by analyzing the "GetRequest" output that is also included in the scan:

```
|   GetRequest: 
|     HTTP/1.0 400 Bad Request
|     Accept-Ranges: bytes
|     Content-Length: 232
|     Content-Type: application/xml
|     Server: MinIO
|     Vary: Origin
|     Date: Thu, 06 Apr 2023 06:46:32 GM
```

I can implement this, but I thought I might discuss this first.

**Comment by fmunozs** [comment]
@h0ng10 can you try running the scan with intensity 7 (nmap default)? I reported a similar problem at #298 in my case nmap is able to detect the HTTP with intensity 7, after that the service can be scanned.

**Comment by h0ng10** [comment]
@fmunozs Yes, I already did that, with the same result.

**Comment by maoning** [comment]
@h0ng10 Thanks for reporting this behaviour. I'm aware in certain cases nmap defaults to /etc/services for open ports, but never really investigated thoroughly. What you suggested makes sense, please go ahead with the proposed solution.

**Comment by h0ng10** [comment]
@maoning I didn't use this in the MinIO plugin, it was just an example I had on hand. 
I think it would be worth trying to at least parse the HTTP headers, as this could also possibly identify the service used (as nmap does).

I will first do some nmap scans on various HTTP services (I already have a few candidates in mind that cause problems) and then see what works best.

**Comment by tooryx** [comment]
Hi,

This has been fixed in [7cf51a42](https://github.com/google/tsunami-security-scanner/commit/7cf51a428953bdb7e44a123a2fbdc38b4d726e09) and [55aa6978](https://github.com/google/tsunami-security-scanner/commit/55aa697873ad90af48901a953150924fa42d92ff)

_~tooryx_
